stoppable_workers.go

```diff
@@ -0,0 +1,71 @@
+package utils
+
+import (
+	"context"
+	"sync"
+)
+
+// StoppableWorkers is a collection of goroutines that can be stopped at a later time.
+type StoppableWorkers interface {
+	AddWorkers(...func(context.Context))
+	Stop()
+	Context() context.Context
+}
+
+// stoppableWorkersImpl is the implementation of StoppableWorkers. The linter will complain if you
+// try to make a copy of something that contains a sync.WaitGroup (and returning a value at the end
+// of NewStoppableWorkers() would make a copy of it), so we do everything through the
+// StoppableWorkers interface to avoid making copies (since interfaces do everything by pointer).
+type stoppableWorkersImpl struct {
+	mu                      sync.Mutex
```

Comment:
As mentioned offline, NetCode has been using `sync.RWMutex`es for groups of goroutines. We use these instead of `sync.Mutex` so that additions of new workers do not lock with each other in any way. We want as-fast-as-possible worker addition; so _read-locking_ only around context-checking and waitgroup-addition makes the most sense to us. `Stop` is the thing that should write-lock and "stop the world".